src/pool/srv_target.c

```diff
@@ -1766,7 +1766,21 @@ obj_discard_cb(daos_handle_t ch, vos_iter_entry_t *ent,
 
 	epr.epr_hi = arg->tgt_discard->epoch;
 	epr.epr_lo = 0;
-	rc = vos_discard(param->ip_hdl, &ent->ie_oid, &epr, NULL, NULL);
+	do {
+		/* Inform the iterator and delete the object */
+		*acts |= VOS_ITER_CB_DELETE;
+		rc = vos_discard(param->ip_hdl, &ent->ie_oid, &epr, NULL, NULL);
+		if (rc != -DER_BUSY && rc != -DER_INPROGRESS)
+			break;
+
+		D_DEBUG(DB_REBUILD, "retry by "DF_RC"/"DF_UOID"\n",
+			DP_RC(rc), DP_UOID(ent->ie_oid));
+		/* Busy - inform iterator and yield */
+		*acts |= VOS_ITER_CB_YIELD;
```

Comment:
VOS iterator is able to detect if yield happen in iter callback, this flag isn't necessary anymore.